language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -92,33 +97,45 @@ defaultTable = do
 
     pure $ BuiltinTable tyTable termTable
 
--- | Type-check a PLC program, returning the inferred normalized type.
-typecheckProgram :: (MonadError (Error a) m, MonadQuote m) => Natural -> Program TyNameWithKind NameWithType a -> m (NormalizedType TyNameWithKind ())
-typecheckProgram n (Program _ _ t) = typecheckTerm n t
-
--- | Type-check a PLC term, returning the inferred normalized type.
-typecheckTerm :: (MonadError (Error a) m, MonadQuote m) => Natural -> Term TyNameWithKind NameWithType a -> m (NormalizedType TyNameWithKind ())
-typecheckTerm n t = convertErrors asError $ runTypeCheckM n (typeOf t)
-
--- | Kind-check a PLC term.
-kindCheck :: (MonadError (Error a) m, MonadQuote m) => Natural -> Type TyNameWithKind a -> m (Kind ())
-kindCheck n t = convertErrors asError $ runTypeCheckM n (kindOf t)
+-- | Type-check a PLC program, returning a normalized type.
+typecheckProgram :: (MonadError (Error a) m, MonadQuote m)
+                 => Natural
+                 -> Bool
+                 -> Program TyNameWithKind NameWithType a
+                 -> m (NormalizedType TyNameWithKind ())
+typecheckProgram n norm (Program _ _ t) = typecheckTerm n norm t
+
+-- | Type-check a PLC term, returning a normalized type.
+typecheckTerm :: (MonadError (Error a) m, MonadQuote m)
+              => Natural
+              -> Bool
+              -> Term TyNameWithKind NameWithType a
+              -> m (NormalizedType TyNameWithKind ())
+typecheckTerm n norm t = convertErrors asError $ runTypeCheckM n norm (typeOf t)
+
+-- | Kind-check a PLC type.
+kindCheck :: (MonadError (Error a) m, MonadQuote m)
+          => Natural
+          -> Bool
+          -> Type TyNameWithKind a
+          -> m (Kind ())
+kindCheck n norm t = convertErrors asError $ runTypeCheckM n norm (kindOf t)
 
 -- | Run the type checker with a default context.
-runTypeCheckM :: (MonadError (TypeError a) m, MonadQuote m)
-              => Natural -- ^ Amount of gas to provide typechecker
+runTypeCheckM :: Natural -- ^ Amount of gas to provide typechecker
+              -> Bool -- ^ Whether to normalize types
```

Comment:
This comment is provided here, but not in any of the functions above. Just define an actual flag, really, that would make type signatures and function calls much more nice and descriptive.